pycargoebuild/cargo.py

```diff
@@ -77,24 +77,52 @@ def get_crates(f: typing.BinaryIO) -> typing.Generator[Crate, None, None]:
                                f"{p!r}") from e
 
 
-def get_package_metadata(f: typing.BinaryIO) -> PackageMetadata:
-    """Read package from the open ``Cargo.toml`` file"""
-    cargo_toml = tomllib.load(f)
-
-    if "package" not in cargo_toml and "workspace" in cargo_toml:
-        raise RuntimeError(
-            "Specified directory seems to be a workspace root, please run "
-            "pycargoebuild on one of its members instead: "
-            f"{' '.join(cargo_toml['workspace']['members'])}")
-
-    pkg_meta = cargo_toml["package"]
-    pkg_license = pkg_meta.get("license")
-    if pkg_license is not None:
-        pkg_license = cargo_to_spdx(pkg_license)
-    return PackageMetadata(
-        name=pkg_meta["name"],
-        version=pkg_meta["version"],
-        license=pkg_license,
-        license_file=pkg_meta.get("license-file"),
-        description=pkg_meta.get("description"),
-        homepage=pkg_meta.get("homepage"))
+class CargoToml:
+    def __init__(self, f: typing.BinaryIO):
+        cargo_toml = tomllib.load(f)
+
+        if "package" not in cargo_toml and "workspace" in cargo_toml:
+            raise RuntimeError(
+                "Specified directory seems to be a workspace root, please run "
+                "pycargoebuild on one of its members instead: "
+                f"{' '.join(cargo_toml['workspace']['members'])}")
+
+        self.workspace_pkg_meta = (cargo_toml.get("workspace", {})
+                                             .get("package", {}))
+        self.pkg_meta = cargo_toml["package"]
+
+    def _get_key(self, key: str, pkg_meta: dict, backup_pkg_meta: dict,
+                 required: bool = False) -> typing.Optional[str]:
+        value = pkg_meta.get(key)
+        if value is None:
+            if required:
+                raise KeyError(f"Metadata key not found: '{key}'")
+            return None
+
+        if isinstance(value, str):
+            return value
+        if isinstance(value, dict):
+            if value.get("workspace") is True:
+                return self._get_key(key, backup_pkg_meta, {}, required)
+        raise ValueError(f"Cannot get metadata key: '{key}'")
+
+    def get_key(self, key: str,
```

Comment:
But… why do you need two functions? oO

Reply:
I can't think of an alternative solution x_x